lib/cdo/mailjet.rb

```diff
@@ -0,0 +1,110 @@
+require 'mailjet'
+require 'honeybadger/ruby'
+require_relative './shared_constants/mailjet_constants'
+
+module MailJet
+  include MailJetConstants
+
+  API_KEY = CDO.try(:mailjet_api_key).freeze
+  SECRET_KEY = CDO.try(:mailjet_secret_key).freeze
+
+  def self.enabled?
+    return false unless DCDO.get('use_mailjet', false)
+    !API_KEY.nil? && !SECRET_KEY.nil?
+  end
+
+  def self.subaccount
+    case Rails.env
+    when 'production'
+      'production'
+    when 'staging'
+      'staging'
+    else
+      'development'
+    end
+  end
+
+  def self.create_contact_and_send_welcome_email(user)
+    return unless enabled?
+
+    return if user&.id.nil?
+    return unless user.teacher?
+
+    create_contact(user.email, user.name, user.created_at.to_datetime.rfc3339)
+    send_template_email(
+      user.email,
+      EMAILS[:welcome][:from_address],
+      user.name,
+      EMAILS[:welcome][:from_name],
+      EMAILS[:welcome][:template_id][subaccount.to_sym]
+    )
+  end
+
+  def self.create_contact(email, name, sign_up_date)
+    return unless enabled?
+
+    Mailjet.configure do |config|
+      config.api_key = API_KEY
+      config.secret_key = SECRET_KEY
+      config.api_version = "v3"
+    end
+
+    # If a contact already exists, Mailjet will raise an error.
+    # This shouldn't happen as user emails should be unique, but
+    # we don't want to block sign up if it does.
+    begin
+      Mailjet::Contact.create(
+        is_excluded_from_campaigns: true,
+        email: email,
+        name: name
+      )
+    rescue => exception
+      Honeybadger.notify(exception)
+    end
+
+    # Most likely, the above would fail if a contact already exists.
+    # In that case, we want to update the contact with the sign up date.
+    # However, in the case of a different error, we want to notify Honeybadger,
+    # but not block sign up.
+    begin
+      contact = Mailjet::Contactdata.find(email)
+      contact.update_attributes(
+        data: [
+          {
+            name: 'sign_up_date',
+            value: sign_up_date
+          }
+        ]
+      )
+    rescue => exception
+      Honeybadger.notify(exception)
+    end
+  end
```

Comment:
Right now this function always makes three API calls. Can we reduce that to one or two calls by rearranging it?


* FIND contact by email
* if found, UPDATE the sign_up_date _if necessary_
* else CREATE with sign_up_date attributes

I think that would also make it more straightforward to add test cases for the various code paths.

Reply:
Additionally, this function is squashing _all_ errors. That seems undesireable and I can think of a few scenarios where you definitely don't want to be calling `send_template_email` if `create_contact` fails.

Reply:
> CREATE with sign_up_date attributes

This always requires two API calls because properties have to be updated through a [separate API](https://dev.mailjet.com/email/reference/contacts/contact-properties/).

I think you've got a good point about finding the contact first then branching though -- I'll update.

Reply:
Ah, that's frustrating. In that case perhaps the create doing double duty as an existence check is fine.

Reply:
Refactored and it's much cleaner! I ended up removing the error squashing in general, but I'm very open to discussion there. In particular, moving the call to find the contact up means that the primary error case disappeared, so I'm less attached to squashing the errors. There's a part of me that's considering adding them back, however, in case MailJet is down or something. Thoughts?

Reply:
Depends if we're going to do anything about it. For a general wrapper, I think it makes sense to raise an exception if `create_contact` fails to create a contact. Catching that exception is the responsibility of the caller, and it can decide whether to retry, fail, or move on. Unless there's a specific exception category that this wrapper should be responsible for resolving (for example, I've worked with API's that have rate limits that were resolved by trying again one more time right now)

Reply:
Makes sense! And thinking through it, I don't want to add any exception handling here. When I hook this up into the actual user sign up flow, I might, but that is coming in a future PR.